PowerShellTools/ServiceManagement/PowershellHostProcessFactory.cs

```diff
@@ -0,0 +1,139 @@
+        /// </summary>
+        internal static PowershellHostProcess HostProcess { get; set; }
+
+        /// <summary>
+        /// In case we need to change host process at some point. What we need to do is signal the termination of the process and then create a new one. 
+        /// </summary>
+        internal static void SignalProcessTerminated()
+        {
+            lock (_syncObject)
+            {
+                LazyCreatePowershellHostProcess();
+            }
+        }
+
+        internal static PowershellHostProcess EnsurePowershellHostProcess()
+        {
+            return _powershellHostProcess.Value;
+        }
+
+        private static PowershellHostProcess CreatePowershellHostProcess()
+        {
+            Process powershellHostProcess = new Process();
+            string hostProcessReadyEventName = Constants.ReadyEventPrefix + Guid.NewGuid();
+            Guid endPointGuid = Guid.NewGuid();
+
+            string exeName = Constants.PowershellHostExeName;
+            string currentPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            string path = Path.Combine(currentPath, exeName);
+            string hostArgs = String.Format(CultureInfo.InvariantCulture,
+                                            "{0}{1} {2}{3} {4}{5}",
+                                            Constants.UniqueEndpointArg, endPointGuid, // For generating a unique endpoint address 
+                                            Constants.VsProcessIdArg, Process.GetCurrentProcess().Id,
+                                            Constants.ReadyEventUniqueNameArg, hostProcessReadyEventName); 
+
+            powershellHostProcess.StartInfo.Arguments = hostArgs;
+            powershellHostProcess.StartInfo.FileName = path;
+            powershellHostProcess.StartInfo.CreateNoWindow = true; 
+            powershellHostProcess.StartInfo.UseShellExecute = true;
+            powershellHostProcess.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
+
+            EventWaitHandle readyEvent = new EventWaitHandle(false, EventResetMode.ManualReset, hostProcessReadyEventName);
+
+            powershellHostProcess.Start();
+            powershellHostProcess.EnableRaisingEvents = true;
+            powershellHostProcess.Exited += PowershellHostProcess_Exited;
+            bool success = readyEvent.WaitOne(Constants.HostProcessStartupTimeout, false);            
+            readyEvent.Close();
+
+            if (!success)
+            {
+                try
+                {
+                    powershellHostProcess.Kill();
+                }
+                catch (Exception)
+                {
+                }
+
+                if (powershellHostProcess != null)
+                {
+                    powershellHostProcess.Dispose();
+                    powershellHostProcess = null;
+                }
+                throw new PowershellHostProcessException(String.Format(CultureInfo.CurrentCulture,

```

Comment:
This is a bit of over-kill to me, it looks more like a regular error handling scenario instead of firing customize exception.